README.md

```diff
@@ -60,7 +60,83 @@ Once you have implemented a lower bound search for the start of the range, imple
+2. We will run the code below if the last index is equal or larger than the first index. 
+3. We end up in this condition if we already reach the very left of array, where last = first = 0. If the value in that index is equal or larger than the number we are looking for, we return 0, otherwise we return 1.
+4. Otherwise, we do the binary search. We check if value in the mid index, is the same or larger than the number we search.
+5. If yes, then we check if the value in the index before the mid is smaller than the number we search. `This part is where it differs from the normal binary search.`
+6. If it also yes, this mid index is indeed the lower bound of the number we are looking for.
+7. Otherwise, we will continue doing a binary search in the  left part of mid index
+8. If the value in mid index we found is smaller than what we are looking for, so it must be in the right side of our mid index.
+
 *Would anything be more difficult if the features covered ranges instead of single nucleotides (like real BED files)? What could go wrong, if anything?*
+We are looking at the chrom_start for the features. This would ot change if the features wasn't SNPs. 
+All features with start within a certain region was extracted, and this would still be possible even if the features wasn't SNPs. 
+If we want the entire feature to be within the range given by the query file we would have to change the extract_features to include the end-position. 
+
+If the features wasn't SNPs we would also have to account for how we wanted them sorted if several features has the same start but different lengths. 
+The obvius would be to sort them with the shortest/them with the lowest chrom_end first. But we would have to include this in the code. 
+And the same thing for merging. 
 
 *We wrote a tool for merging two BED files, but what if we had a bunch of them? What would the complexity be if we merged them in, one at a time? What would the complexity be if we merged all of the files at the same time?*
 
+The complexity of the merge we have made is O(n+m), where n and m are the lenghts of the two files we need to merge. This is because we need to go through each feature individually and copy it to a new merged list. 
+If we were to merge more than two files one by one, it would just be O(n+m+o+p....) where the letters are the lenghts of the files for however many files we want to merge. 
+If we merge lists with lengths n, m and o would the complexity be O(n+m) for the merge of the two first list and O(n+m+o), which boils down to O(n+m+o) (linear time)
+If we merged three lists at one it would still happen in linear time O(n+m+o). 
+Even tough the complexity is the same for both methods every element of the first two lists will be copied twice with the first method and only once for the second method. 
+The number of comparisons on the other hand will be higher for the second method.
```

Comment:
For simplicity, let’s say that all the files have length $m$ and that there are $n$ of them. Merging the first two would take time $2m$ (two files of length $m$). Merging in the third would take $3m$, however (one file of length $2m$ and one of length $m$). The fourth would require time $4m$ (merging one file of length $3m$ with one of length $m$).

In general, we would need the time $m(2+3+4+\cdots+n-1)$ time to merge $n$ files of length $m$ and the sum in that expression (as you know) is in $O(n^2)$, so the running time is $O(mn^2)$ (far from linear).

Concatenating all the files and sorting would instead take $O(nm\log(nm))$, which could be better (depending on $n$ and $m$).

You could also, using divide and conquer, split the input in two $n/2$-length lists of files, merge these recursively, and then merge the results. That would give you the recurrence $T(n,m) = 2T(n/2,m) + O(n\cdot m)$ which is in $O(mn\log(n))$.

If you merge the $n$ files in parallel, so you, in each iteration, find the smallest front element in the $n$ files, you need to copy $nm$ elements, but it takes time to find the smallest. If you do this with a linear scan, it takes $O(n)$ to find the smallest element in each iteration, and you end up back at $O(mn^2)$. You can speed this up (once we have seen heaps in a later lecture) so it takes time $O(\log n)$ to get the smallest element per iteration, and then you are back to $O(mn\log n)$ (just with a completely different approach).

Many roads lead to Rome.